### PR TITLE
Refactor implementation of the total spin observable S2

### PR DESCRIPTION
**Pull request by agran2018**
**Context:**
This is part of the refactoring of the obs module used to build many-body observables.

**Description of the Change:**
The total spin operator S^2 is constructed by calling the function `spin2` as implemented in the obs module.

**Benefits:**
The function to construct the observable is simpler and does not depend on any external-to-pennylane data structure. It returns the observable.

**Possible Drawbacks:**
None

**Related GitHub Issues:**
None

### CODE REVIEW

File: qchem/pennylane_qchem/qchem/obs.py [hunk 1]
```diff
@@ -15,48 +15,47 @@
...
-    r"""Generates the table of matrix elements
-    :math:`\langle \alpha, \beta \vert \hat{s}_1 \cdot \hat{s}_2 \vert \gamma, \delta \rangle`
-    of the two-particle spin operator :math:`\hat{s}_1 \cdot \hat{s}_2`.
+def _spin2_matrix_elements(sz):
+    r"""Builds the table of matrix elements
+    :math:`\langle \bm{\alpha}, \bm{\beta} \vert \hat{s}_1 \cdot \hat{s}_2 \vert
+    \bm{\gamma}, \bm{\delta} \rangle` of the two-particle spin operator
+    :math:`\hat{s}_1 \cdot \hat{s}_2`.
 
     The matrix elements are evaluated using the expression,
```
Comment (josh146) [inline qchem/pennylane_qchem/qchem/obs.py:31]:
```suggestion
    The matrix elements are evaluated using the expression
```

File: qchem/pennylane_qchem/qchem/obs.py [hunk 2]
```diff
@@ -101,98 +95,90 @@ def _spin2_matrix_elements(sz, n_spin_orbs):
...
-        \langle \alpha, \beta \vert \hat{s}_1 \cdot \hat{s}_2
-        \vert \gamma, \delta \rangle ~ \hat{c}_\alpha^\dagger \hat{c}_\beta^\dagger
-        \hat{c}_\gamma \hat{c}_\delta,
+        \hat{S}^2 = \frac{3}{4}N + \sum_{ \bm{\alpha}, \bm{\beta}, \bm{\gamma}, \bm{\delta} }
+        \langle \bm{\alpha}, \bm{\beta} \vert \hat{s}_1 \cdot \hat{s}_2
+        \vert \bm{\gamma}, \bm{\delta} \rangle ~
+        \hat{c}_\bm{\alpha}^\dagger \hat{c}_\bm{\beta}^\dagger
+        \hat{c}_\bm{\gamma} \hat{c}_\bm{\delta},
     
     where the two-particle matrix elements are computedas,
```
Comment (josh146) [inline qchem/pennylane_qchem/qchem/obs.py:111]:
```suggestion
    where the two-particle matrix elements are computed as,
```

File: qchem/pennylane_qchem/qchem/obs.py [hunk 3]
```diff
@@ -101,98 +95,90 @@ def _spin2_matrix_elements(sz, n_spin_orbs):
...
+    + (0.375) [Z3]
+    + (-0.375) [Z2 Z3]
+    + (0.125) [Y0 X1 Y2 X3]
+    + (0.125) [Y0 Y1 X2 X3]
+    + (0.125) [Y0 Y1 Y2 Y3]
+    + (-0.125) [Y0 X1 X2 Y3]
+    + (-0.125) [X0 Y1 Y2 X3]
+    + (0.125) [X0 X1 X2 X3]
+    + (0.125) [X0 X1 Y2 Y3]
+    + (0.125) [X0 Y1 X2 Y3]
```
Comment (josh146) [inline qchem/pennylane_qchem/qchem/obs.py:164]: Nice example!

File: qchem/pennylane_qchem/qchem/obs.py [hunk 4]
```diff
@@ -65,12 +64,7 @@ def _spin2_matrix_elements(sz, n_spin_orbs):
      [ 1.    0.    1.    0.    0.5 ]]
     """
 
-    if sz.size != n_spin_orbs:
-        raise ValueError(
-            "Size of 'sz' must be equal to 'n_spin_orbs'; size got for 'sz' {}".format(sz.size)
-        )
-
-    n = np.arange(n_spin_orbs)
+    n = np.arange(sz.size)
```
Comment (ixfoduap) [inline qchem/pennylane_qchem/qchem/obs.py:67]: Great change!

**Pull request merged**